### PR TITLE
adding parser support for code keyword for data location

### PR DESCRIPTION
**Pull request by nishant-sachdeva**
closes https://github.com/ethereum/solidity/issues/13368

### CODE REVIEW

File: libsolidity/parsing/Parser.cpp [hunk 1]
```diff
@@ -811,6 +811,14 @@ ASTPointer<VariableDeclaration> Parser::parseVariableDeclaration(
 					case Token::CallData:
 						location = VariableDeclaration::Location::CallData;
 						break;
+					case Token::Code:
+						location = VariableDeclaration::Location::Code;
+						syntaxError(
```
Comment (Marenz) [inline libsolidity/parsing/Parser.cpp]: I think an `solUnimplementedError` (or something like that, not sure what the exact name is) would be desired here

**Review by Marenz — CHANGES_REQUESTED** [review]
Looking at the CI there seem to be a few more places that require changes, mostly switch cases that don't handle `code` yet.

Overall the approach looks fine.

**Comment by Marenz** [comment]
Hmm there are a few CI failures that are fixed in `develop`, so we probably need to merge current `develop` into `breaking` after all.

**Comment by matheusaaguiar** [comment]
This change is also affecting Yul. The following minimal example:
```
object "runtime" {
    code {

    }
}
```
compiled with `./solc --strict-assembly example.yul` will result in the error:
```Error: Expected keyword "code".```

File: libsolidity/parsing/Parser.cpp [hunk 2]
```diff
@@ -811,6 +811,9 @@ ASTPointer<VariableDeclaration> Parser::parseVariableDeclaration(
 					case Token::CallData:
 						location = VariableDeclaration::Location::CallData;
 						break;
+					case Token::Code:
+						solUnimplemented("Code keyword as a data location specifier has not been implemented yet.");
```
Comment (matheusaaguiar) [inline libsolidity/parsing/Parser.cpp:815]: I was checking and for this point to be reached, `TokenTraits::isLocationSpecifier` needs to be modified to recognize `Token::Code`. Also, this throws an exception and I understood from #13368 that it should be a syntax error.

Comment (Marenz) [reply libsolidity/parsing/Parser.cpp:815]: My bad advice then, I thought we wanted to use `solUnimplemented()`

**Comment by ekpyron** [comment]
I'm actually not quite sure how to react to this... did any of you actually read the issue?
The task is to add the keyword **and** to actually parse it as a data location, while making any use of the keyword a syntax error. As in an error in ``SyntaxChecker``.
While, technically, this PR is an initial step towards that, this by no means solves the issue.
Also the interactions with Yul should have become apparent with a single local CI run (the issue can be fixed by a minor adjustment of ``ObjectParser::parseCode``).

I also generally don't think it's beneficial to keep draft PRs in a state like this open. So please make sure you understand the scope of the underlying issue, if necessary ask for clarification on it, and reopen a proper PR actually addressing the issue.

**Pull request closed without merging**